src/Parser.js

```diff
@@ -28,236 +33,199 @@ module.exports = class Parser {
-        }
-        case 'html': {
-          out += renderer.html(token.text);
-          break;
-        }
-        case 'link': {
-          out += renderer.link(token.href, token.title, this.parseInline(token.tokens, renderer));
-          break;
-        }
-        case 'image': {
-          out += renderer.image(token.href, token.title, token.text);
-          break;
-        }
-        case 'strong': {
-          out += renderer.strong(this.parseInline(token.tokens, renderer));
-          break;
-        }
-        case 'em': {
-          out += renderer.em(this.parseInline(token.tokens, renderer));
-          break;
-        }
-        case 'codespan': {
-          out += renderer.codespan(token.text);
-          break;
-        }
-        case 'br': {
-          out += renderer.br();
-          break;
-        }
-        case 'del': {
-          out += renderer.del(this.parseInline(token.tokens, renderer));
-          break;
-        }
-        case 'text': {
-          out += renderer.text(token.text);
-          break;
-        }
-        default: {
-          const errMsg = 'Token with "' + token.type + '" type was not found.';
-          if (this.options.silent) {
-            console.error(errMsg);
-            return;
-          } else {
-            throw new Error(errMsg);
+  blockquote(token, ctx) {
+    return ctx.renderer.blockquote(this.parse(token.tokens, ctx));
+  }
+
+  list(token, ctx) {
+    const { ordered, start, items, loose } = token;
+    if (loose) ctx = this.getLooseContext(ctx);
+
+    let body = '';
+    const l = items.length;
+    for (let i = 0; i < l; i++) {
+      body += this.listitem(items[i], ctx);
+    }
+
+    return ctx.renderer.list(body, ordered, start);
+  }
+
+  listitem(item, ctx) {
+    const { checked, task, tokens } = item;
+    const loose = ctx.loose;
+
+    let body = '';
+    if (task) {
+      const checkbox = ctx.renderer.checkbox(checked);
+      if (loose) {
+        if (tokens.length > 0 && tokens[0].type === 'text') {
+          tokens[0].text = checkbox + ' ' + tokens[0].text;
+          if (tokens[0].tokens && tokens[0].tokens.length > 0 && tokens[0].tokens[0].type === 'text') {
+            tokens[0].tokens[0].text = checkbox + ' ' + tokens[0].tokens[0].text;
           }
+        } else {
+          tokens.unshift({ type: 'text', text: checkbox });
         }
+      } else {
+        body += checkbox;
       }
     }
-    return out;
+
+    body += this.parse(tokens, ctx);
+
+    // Regression/placeholder, TODO revise logic. See comment on `.text()`.
+    if (loose) body = ctx.renderer.paragraph(body);
+
+    return ctx.renderer.listitem(body, task, checked);
+  }
+
+  // TODO parse inline content if parameter markdown=1
+  html(token, ctx) {
+    return ctx.renderer.html(token.text);
+  }
+
+  paragraph(token, ctx) {
+    return ctx.renderer.paragraph(this.parse(token.tokens, ctx));
+  }
+
+  // Regression; missing features:
+  //   * Join adjacent text tokens with '\n'.
```

Comment:
Could we add `tokens` and `currentIndex` to the context object so the `text` function could combine tokens the same way it does in the switch statement?

Reply:
Unless I misunderstood, sounds like it involves mutating the context. Parsing/compiling any token, even `text`, is potentially recursive, with nested parse loops. To avoid trampling each other's `tokens` and `currentIndex` in the context, parse loops would have to either make new context objects (slow), or stash some contextual values in the stack frame of the current parse loop, and restore them before returning from the loop (shouldn't affect performance, but looks dirty). This also creates a "communication channel" between a parse loop and the methods it's calling, though a shared mutable object. This could work, but a cleaner approach would be much appreciated.

I admit this is a self-inflicted problem. The "old" loop had the power to handle this case and worked just fine. But I also feel that making the parse loop really dumb, where it just calls one method per token, "buys" us nice extensibility through method overrides, and is at least worth exploring.

I don't entirely understand the reasoning behind this special case. Why would there be adjacent text tokens? Is this currently generated anywhere inside Marked, or supported for outside customization? Could this be handled better at the Lexer level? Could it be handled better through `text.tokens`?

Reply:
> "buys" us nice extensibility through method overrides

If the goal is just to have extensibility it seems like it would be easier to just have it check for function in the `default` case.

> but a cleaner approach would be much appreciated

Clean code isn't really what marked is aiming for. Clean code often comes with slow performance. We try to make it as clean as possible without sacrificing performance at all. The switch statement may be ugly but it works and is as fast as it can be. Perhaps the goal of a "dumb" parse loop is the wrong goal.